### ISSUE TITLE
loss of type safety when using function overloads with callbacks

### ISSUE THREAD

**Issue by rkirov** [Microsoft/TypeScript#13430]
**TypeScript Version:**  2.1.4
**Code**
```ts
declare function it(f: () => void): number;
declare function it(f: (x: string) => void): string;

let r = it((x) => {x});
```

**Expected behavior:**
The inferred type of `x` inside the closure should be `string`.

**Actual behavior:**
The inferred type of `x` is `any`. Moreover, this triggers noImplicitAny errors.

It appears that the first overload somehow throws off the type inference (no issue without overloads). However, the resulting type of `r` is `string` thus the second overload was ultimately chosen. Despite that, the type of `x` is not inferred to match the one in the overload.

This example worked as expected in TS 2.0.x so it appears to be a regression.

//cc @calebegg

**Comment by aluanhaddad** [comment]
```ts
declare function it(f: (x: string) => void): string;
declare function it(f: () => void): number;
```
The more specific overload needs to be declared first.

**Comment by rkirov** [comment]
@aluanhaddad note the return type of `r` means the second overload was chosen. Also this used to work in 2.0.x, any idea what changed?

**Comment by aluanhaddad** [comment]
>note the return type of r means the second overload was chosen

I'm not sure what you mean, the correct overload is chosen if ordered as shown
```ts
declare function it(f: (x: string) => void): string;
declare function it(f: () => void): number;

let r = it((x) => {x}); // string
```

> Also this used to work in 2.0.x, any idea what changed?

I believe this was the result of a fix a bug that was fixed as this behavior is intentional, see https://github.com/Microsoft/TypeScript/issues/10645#issuecomment-253582156.

**Comment by rkirov** [comment]
I meant is that if you go -

http://www.typescriptlang.org/play/#src=declare%20function%20it(f%3A%20()%20%3D%3E%20void)%3A%20number%3B%0Adeclare%20function%20it(f%3A%20(x%3A%20string)%20%3D%3E%20void)%3A%20string%3B%0A%0Alet%20r%20%3D%20it((x)%20%3D%3E%20%7Bx%7D)%3B

and hover over `r` you see it is `string`, while if you hover over `x`- it is `any`.

**Comment by mhegazy** [comment]
This should be now flagged as an implicit any under `--noImplicitAny`. This is really a side effect of how the compiler does overload resolution with respect to context sensitive lambdas. Overloads are tried in order, and context sensitive lambdas need a contextual type, the compiler will apply the one from the first overload, but once that happens it can not un-apply it, since contextual types have implications on the rest of the expressions inside the function.
